jit-compiler/src/codegen.rs

```diff
@@ -416,8 +450,37 @@ fn check_pattern(value_name: &str, pattern: &Pattern) -> Result<(String, String)
             )
         }
         Pattern::Variable(_, var) => (var.to_string(), format!("Some({value_name}.clone())")),
-        Pattern::Enum(..) => {
-            return Err(format!("Enums as patterns not yet implemented: {pattern}"));
+        Pattern::Enum(_, symbol, None) => (
+            "_".to_string(),
+            format!(
+                "(matches!({value_name}, {}).then_some(()))",
+                escape_enum_variant(symbol)
+            ),
+        ),
+        Pattern::Enum(_, symbol, Some(items)) => {
+            // We first match the enum variant and bind all items to variables and
+            // the recursively match the items, even if they are catch-all.
+            let mut vars = vec![];
+            let item_name = |i| format!("item__{i}");
+            let inner_code = items
+                .iter()
+                .enumerate()
+                .map(|(i, item)| {
+                    let (v, code) = check_pattern(&item_name(i), item)?;
+                    vars.push(v);
+                    Ok(format!("({code})?"))
+                })
+                .collect::<Result<Vec<_>, String>>()?
+                .join(", ");
+
+            (
+                vars.join(", "),
+                format!(
+                    "(|| if let {}({}) = ({value_name}).clone() {{ Some({inner_code}) }} else {{ None }})()",
```

Comment:
Why this must be a closure call?

Reply:
So that the inner code can use the `?` operator.